_posts/2022-10-27-Adding-New-Distributions-to-OpenSearch-Project.md

```diff
@@ -0,0 +1,411 @@
+
+```python
+from build_workflow.build_args import BuildArgs
+......
+args = BuildArgs()
+```
+
+Next, we add the following code in [src/build_workflow/build_args.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-f5a380768626613f257a019bff13f8aec914e2b2983cea9d7a4552436eba4a45) using the `argparse` module.
+
+```python
+# tar and zip already existed prior to our change
+# we will keep them so that olded functions are not broken
+
+SUPPORTED_DISTRIBUTIONS = ["tar", "zip", "rpm"]
+......
+parser.add_argument(
+    "-d",
+    "--distribution",
+    type=str,
+    choices=self.SUPPORTED_DISTRIBUTIONS,
+    help="Distribution to build.",
+    default=None,
+    dest="distribution"
+)
+```
+
+Other files will need updates to support this new parameter. Examples include [src/build_workflow/build_target.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-22bbfc8727ab963ed470460b1996e12fbb9974a4caafb113fd2319ffa08a0a46) and [src/run_build.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-db6a1b8175460e03234674e4c8951029102a373968c4f6cc87e01b17301c2ea8).
+
+When running build, we use an input manifest file (for this blog, it's `manifest/1.3.2/opensearch-1.3.2.yml`). This file consists of the components we want to build (the OpenSearch engine and related plugins). To learn more about the manifest files, read the code in [src/manifests](https://github.com/opensearch-project/opensearch-build/tree/1.3.2/src/manifests) for more information. Two other manifest files (both named `manifest.yml` ) are generated after `build_workflow` and `assemble_workflow` record the build and assemble results. The two `manifest.yml` files are referred to as “build manifest” and “bundle manifest”. Sometimes, “bundle manifest” is also referred to as “dist manifest”. 
+
+To add the distribution key-value pair in the build manifest so the `assemble_workflow` can use the build manifest to assemble the actual artifact in the given distribution format, we update [src/build_workflow/build_recorder.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-124059c5fa06b48851cc78f9260a851e37dec9dbbb3e9ea7cde945ad7b159c75) and [src/manifests/build_manifest.py](https://github.com/opensearch-project/opensearch-build/pull/1629/files#diff-728703e829193efc4033a1975aee498a8752fcd2999254326c08865a4af19992).
+
+In `build_recorder.py`, we define the default value `distribution` as `tar`:
+
+```python
+ self.data["build"]["distribution"] = target.distribution if target.distribution else "tar"
+```
+
+In `build_manifest.py`, `distribution` is `None` so that older `manifest.yml` without the `distribution` key-value pair is compatible with the new code:
```

Comment:
addressed to:
```
In `build_manifest.py`, `distribution` is `None` so that the older `manifest.yml` files from previous releases without the `distribution` key-value pair are compatible with the new code:
```